src/storage/raw/raw_mvcc.rs

```diff
@@ -0,0 +1,215 @@
+use tikv_util::codec::number::U64_SIZE;
+use txn_types::{Key, TimeStamp, Value};
+
+#[derive(Clone)]
+pub struct RawMvccSnapshot<S: Snapshot> {
+    snap: S,
+}
+
+impl<S: Snapshot> RawMvccSnapshot<S> {
+    pub fn from_snapshot(snap: S) -> Self {
+        RawMvccSnapshot { snap }
+    }
+
+    pub fn seek_first_key_value_cf(
+        &self,
+        cf: Option<CfName>,
+        opts: Option<ReadOptions>,
+        key: &Key,
+    ) -> Result<Option<Value>> {
+        let mut iter_opt = IterOptions::default();
+        iter_opt.set_fill_cache(opts.map_or(true, |v| v.fill_cache()));
+        iter_opt.use_prefix_seek();
+        iter_opt.set_prefix_same_as_start(true);
+        let end_key = key.clone().append_ts(TimeStamp::zero());
+        iter_opt.set_upper_bound(end_key.as_encoded(), DATA_KEY_PREFIX_LEN);
+        let mut iter = match cf {
+            Some(cf_name) => self.iter_cf(cf_name, iter_opt)?,
+            None => self.iter(iter_opt)?,
+        };
+        if iter.seek(key)? {
+            Ok(Some(iter.value().to_owned()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<S: Snapshot> Snapshot for RawMvccSnapshot<S> {
+    type Iter = RawMvccIterator<S::Iter>;
+    type Ext<'a>
+    where
+        S: 'a,
+    = S::Ext<'a>;
+
+    fn get(&self, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(None, None, key)
+    }
+
+    fn get_cf(&self, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(Some(cf), None, key)
+    }
+
+    fn get_cf_opt(&self, opts: ReadOptions, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(Some(cf), Some(opts), key)
+    }
+
+    fn iter(&self, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter(iter_opt)?))
+    }
+
+    fn iter_cf(&self, cf: CfName, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter_cf(cf, iter_opt)?))
+    }
+
+    #[inline]
+    fn lower_bound(&self) -> Option<&[u8]> {
+        self.snap.lower_bound()
+    }
+
+    #[inline]
+    fn upper_bound(&self) -> Option<&[u8]> {
+        self.snap.upper_bound()
+    }
+
+    fn ext(&self) -> S::Ext<'_> {
+        self.snap.ext()
+    }
+}
+
+pub struct RawMvccIterator<I: Iterator> {
+    inner: I,
+    cur_key: Option<Vec<u8>>,
+    cur_value: Option<Vec<u8>>,
+    is_valid: Option<bool>,
+}
+
+fn is_user_key_eq(left: &[u8], right: &[u8]) -> bool {
+    let len = left.len();
+    if len != right.len() {
+        return false;
+    }
+    Key::is_user_key_eq(left, &right[..len - U64_SIZE])
+}
+
+impl<I: Iterator> RawMvccIterator<I> {
+    fn new(inner: I) -> Self {
+        RawMvccIterator {
+            inner,
+            cur_key: None,
+            cur_value: None,
+            is_valid: None,
+        }
+    }
+
+    fn update_cur_kv(&mut self) {
+        if let Some(ref mut key) = self.cur_key {
+            key.clear();
```

Comment:
It may need to be shrink if key becomes too long.

Reply:
good proposal. how about shrinking the vector when capacity is 2 times than the key size?

Reply:
There should be a hard limit. For example, it's meaningless to shrink a 512 buffer.

Reply:
ok